docs/weave/design/01-extensions.md

```diff
@@ -0,0 +1,45 @@
+
+Note that even pieces as fundamental as [signature validation](https://github.com/iov-one/weave/tree/master/x/sigs) or [isolating failed transactions](https://github.com/iov-one/weave/blob/master/x/utils/savepoint.go) are implemented as importable modules and wired up together when you construct the application.
+
+## Extension Functionality
+
+Most code to extend the basic framework is packaged as *extensions* and stored in packages under the x directory. One can ignore all code under x and still build a complete weave-compatible application, these are just some pieces that many chains would like to reuse. You can import these extensions, or write you own, with the same power and functionality.
+
+When you write a weave-based application (covered in the next section), you will likely want to create a few new extensions to add new functionality tied into why your chain is unique. What types of behavior can you customize?
+
+- Handler - process transactions, maintain local state, control state transitions. Sending coins is an example of a Handler
+- Decorator (aka Middleware) - do some pre-processing and update the context with information to influence eventual Handler. Signature validation is an example of a Decorator.
+- Initializer - Provide a function to initialize the data store based on a section of the app\_state in the genesis file
+
+\* Handle ABCI calls - `app.BaseApp` implements `abci.Application` and you can wrap it with a pure ABCI Middleware to do things like record timing information on the `Commit` or `Info` calls, without forking the code.
+
+## App Framework
+
+If `weave` allows you to customize everything, what does it provide?
+
+### ORM
+
+The `orm` package wraps up the merkle tree, provable kv-store provided by [tendermint iavl](https://github.com/tendermint/iavl) and adds convenient features on top, like `CacheWrap` to isolate read/writes of a transaction before deciding to Write or Discard, and provides type-safe data storage and secondary indexes if you write to an `orm.Bucket`. In fact, even if you want to build your own framework from scratch, take a look about using `orm` and `iavl` together to provide storage
+
+### ABCI Adapter
+
+The `app` package builds on top of the `orm` to provide default implementations for many of the abci calls, and parses the other ones to allow us to handle requests easier with internal functions. One one hand, it adapts the format, so we can do things like locally return changes to the validator set during `DeliverTx` (when we calculate it), but return the final change on `EndBlock` (when tendermint expects the response). It also handles routing transactions and queries to various modules rather than one monolith. This package demonstrates where most of the main interfaces are used for.
```

Comment:
```suggestion
The `app` package builds on top of the `orm` to provide default implementations for many of the abci calls, and parses the other ones to allow us to handle requests easier with internal functions. One one hand, it adapts the format, so we can do things like locally return changes to the validator set during `DeliverTx` (when we calculate it), but return the final change on `EndBlock` (when Tendermint expects the response). It also handles routing transactions and queries to various modules rather than one monolith. This package demonstrates where most of the main interfaces are used for.
```